### PR TITLE
[Op] Migration: Gradients for some operators

### PR DESCRIPTION
**Pull request by SiriusNEO**
This PR migrates https://github.com/mlc-ai/relax/pull/46 to new struct info infra, as part of our AD migration.

Because we need do numerical testing for gradients, this PR depends on the operator legalizer https://github.com/mlc-ai/relax/pull/96. Also because the original version of legalizer did not handle the negative indexing case of `relax.mean`, this PR fixes it.

To lower `collapse_sum_to`, `collapse_sum_like` properly, this PR migrates a previous patch https://github.com/mlc-ai/relax/pull/43 which introduces `collapse_sum` in topi. Now we can remove the skip marker in the legalizer test for `collapse_sum_to` and `collapse_sum_like`.

The gradients of `cross_entropy` and `softmax_cross_entropy` are removed. And the former will be added back and adjust to new `cross_entropy` introduced in https://github.com/mlc-ai/relax/pull/96.

Further plan in this PR:
- [x] Add gradients for `log_softmax` and `nll_loss` once https://github.com/mlc-ai/relax/pull/94 is merged.
- [x] Gradients for some tuple related operators such as `split` and `concat`. It can help us to test the correctness of AD when there are Tuple-I/O operators.
- (Not in this PR) "Undefined Gradient" representation. As we know, the gradients of some operators w.r.t. specified inputs are undefined or meaningless, such as the partial gradient of `indices` in `take(x, indices)`. Relay directly uses `zeros_like` in this case as it won't affect gradient propagation. Another choice is to introduce a dummy Expr named `UndefinedGradient` to represent it. How do we handle this case in relax?

### CODE REVIEW

File: python/tvm/relax/op/gradient.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    """Gradient of softmax."""
+    return [multiply(subtract(grad, _sum(multiply(grad, orig), orig.attrs.axis, True)), orig)]
+
+
+@register_gradient("relax.sigmoid")
+def sigmoid_grad(orig: Call, grad: Var):
+    """Gradient of sigmoid.
+
+        y = relax.sigmoid(x)
+
```
Comment (Ubospica) [inline python/tvm/relax/op/gradient.py]: Seems these annotations are useless? We do not specify what is y and what is x, and people would assume that `orig = R.sigmoid(orig.args[0])` because of the convention of `@register_gradient`. 

Or we could add more math annotations here

File: python/tvm/relax/op/gradient.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        )
+    ]
+
+
+@register_gradient("relax.tanh")
+def tanh_grad(orig: Call, grad: Var):
+    """Gradient of tanh.
+
+        y = relax.tanh(x)
+
```
Comment (Ubospica) [inline python/tvm/relax/op/gradient.py]: Ditto

File: python/tvm/relax/op/gradient.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    x_ones = ones(x.struct_info.shape, x.struct_info.dtype)
+    return [where(less(x, x_zeros), x_zeros, multiply(x_ones, grad))]
+
+
+@register_gradient("relax.matmul")
+def matmul_grad(orig: Call, grad: Var):
+    """Gradients for matmul.
+
+        c = relax.matmul(a, b)
+
```
Comment (Ubospica) [inline python/tvm/relax/op/gradient.py]: Ditto

File: python/tvm/relax/op/gradient.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    axes = orig.attrs["axes"]
+    if axes:
+        return [permute_dims(grad, axes=axes)]
+    else:
+        return [permute_dims(grad)]
+
+
+@register_gradient("relax.nn.relu")
+def relu_grad(orig: Call, grad: Var):
+    """Returns grad * (select(x < 0, 0, 1))."""
```
Comment (Ubospica) [inline python/tvm/relax/op/gradient.py]: Shall we use `where` instead of select?

File: python/tvm/relax/op/gradient.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        collapse_sum_to(multiply(grad, y), x.struct_info.shape),
+        collapse_sum_to(multiply(grad, x), y.struct_info.shape),
+    ]
+
+
+@register_gradient("relax.permute_dims")
+def permute_dims_grad(orig: Call, grad: Var):
+    """Returns grad transposed over the complement of original permute_dims axes."""
+    axes = orig.attrs["axes"]
+    if axes:
```
Comment (Ubospica) [inline python/tvm/relax/op/gradient.py]: Fixed, just for record

We should use the **inverse** of `axes`, instead of `axes`, as the new axes of the gradient of permute_dims.

File: python/tvm/relax/op/gradient.py (hunk 1 above)
Comment (SiriusNEO) [reply python/tvm/relax/op/gradient.py]: I think we can add some descriptions such as
```
Here x = orig.args[0], y = orig
```
Using `orig.args[0]` directly in the formula makes it too long and hard to read

File: python/tvm/relax/op/base.py [hunk 6]
```diff
@@ -411,3 +411,18 @@ def shape_of(expr: Expr) -> Expr:
         A relax Call, which gets the shape of the input
     """
     return _ffi_api.shape_of(expr)  # type: ignore # pylint: disable=no-member
+
+
+def register_gradient(op_name: str, fgradient: Callable[[Call, Var], Expr] = None, level: int = 10):
```
Comment (Ubospica) [inline python/tvm/relax/op/base.py]:
```suggestion
def register_gradient(op_name: str, fgradient: Callable[[Call, Var], List[Expr]] = None, level: int = 10):
```

**Comment by Ubospica** [comment]
Shall we import `._op_gradient` in `tvm.relax.op.__init__`? If so, we would have operator gradients whereever we use operators.

And with this we don't have to `import tvm.relax.op._op_gradient` in `tests/python/relax/test_op_gradient_numeric.py`.

File: python/tvm/topi/reduction.py [hunk 7]
```diff
@@ -248,3 +248,28 @@ def prod(data, axis=None, keepdims=False):
...
+    ----------
+    data : tvm.te.Tensor
+        The input tensor.
+    shape : Tuple[int]
+        The shape to collapse to.
+    Returns
+    -------
+    ret : tvm.te.Tensor
+        The result tensor after summation.
+    """
```
Comment (Ubospica) [inline python/tvm/topi/reduction.py]: Seems that we have lost empty lines between paragraphs

File: python/tvm/relax/op/_op_gradient.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+
+    return [
+        collapse_sum_to(a_grad, tensor_a.struct_info.shape),
+        collapse_sum_to(b_grad, tensor_b.struct_info.shape),
+    ]
+
+
+@register_gradient("relax.sum")
+def sum_grad(orig: Call, grad: Var):
+    """Gradient of sum."""
```
Comment (Ubospica) [inline python/tvm/relax/op/_op_gradient.py]:
```suggestion
    """Gradient of sum.
    
        y = relax.sum(x, axis, keepdims)
        
    Returns [broadcast_to(y_grad, x.shape)]. If `keepdims=False`, the summed axis will be added back.
    """
```

Seems that we can unify the wording, like "Gradients of " or whatever LOL

File: python/tvm/relax/op/_op_gradient.py [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+    x, y = orig.args
+    return [
+        collapse_sum_to(multiply(grad, y), x.struct_info.shape),
+        collapse_sum_to(multiply(grad, x), y.struct_info.shape),
+    ]
+
+
+@register_gradient("relax.permute_dims")
+def permute_dims_grad(orig: Call, grad: Var):
+    """Returns grad transposed over the complement of original permute_dims axes."""
```
Comment (Ubospica) [inline python/tvm/relax/op/_op_gradient.py]:
```suggestion
    """Gradients for permute_dims.
    
        y = relax.permute_dims(x, axes)
    
    Returns grad transposed over the **inverse permutation** of the original permute_dims axes."""
```

**Review by Ubospica — DISMISSED** [review]

**Comment by MasterJH5574** [comment]
Hey I notice that this PR actually contains four parts that can be decoupled:
* Unary op: exp
* TOPI support for collapse_sum
* Legalizer support for a few ops
* Gradient for each op.

To ease the effort of future sync, would you mind send the first three parts (with proper unit tests) as separate patches? We can have quick review to merge them.

File: python/tvm/relax/block_builder.py [hunk 10]
```diff
@@ -35,7 +35,7 @@
     Binding,
 )
 from .struct_info import ShapeStructInfo, StructInfo, TensorStructInfo
-from .op.base import call_tir
```
Comment (MasterJH5574) [inline python/tvm/relax/block_builder.py:38]: Just curious, is there any specific reason for doing so?

File: python/tvm/relax/op/_op_gradient.py [hunk 11]
```diff
@@ -0,0 +1,488 @@
...
+        shape = expr.struct_info.shape
+    except Exception as error:
+        raise TVMError(
+            f"Get the shape of {expr} failed. Please normalize it first and ensure it is a Tensor."
+        ) from error
+    return shape
+
+
+def _fit_shape(expr: Expr, expr_shape: ShapeExpr, target: Expr) -> Expr:
+    target_shape = _get_shape(target)
```
Comment (MasterJH5574) [inline python/tvm/relax/op/_op_gradient.py]: Let’s also have some docstring for the helper func

File: python/tvm/relax/op/_op_gradient.py [hunk 12]
```diff
@@ -0,0 +1,488 @@
...
+    negative,
+    exp,
+)
+from .binary import (
+    subtract,
+    multiply,
+    divide,
+    less,
+)
+from .statistical import sum as _sum
```
Comment (MasterJH5574) [inline python/tvm/relax/op/_op_gradient.py]: I think it’s fine to disable “redefined builtin” for pylint and directly import sum here

File: python/tvm/relax/op/_op_gradient.py [hunk 13]
```diff
@@ -0,0 +1,488 @@
...
+    def _check_shape_equal():
+        expr_sinfo = expr_shape.struct_info
+        target_sinfo = target_shape.struct_info
+        if len(expr_sinfo.values) != len(target_sinfo.values):
+            return False
+        analyzer = Analyzer()
+        for i, field in enumerate(expr_sinfo.values):
+            if not analyzer.can_prove_equal(field, target_sinfo.values[i]):
+                return False
+        return True
```
Comment (MasterJH5574) [inline python/tvm/relax/op/_op_gradient.py]: We have similar shape analysis function
https://github.com/mlc-ai/relax/blob/760dbe8a71361654fa64fa5f2d8a31f777f7d1ac/include/tvm/relax/analysis.h#L43-L70

But it isn’t exposed to FFI and Python side, and it also assumes the input Expr are ShapeExpr. You can add some new function to .h/.cc/.py accordingly for your case.

Comment (SiriusNEO) [reply python/tvm/relax/op/_op_gradient.py]: Let me expose it to FFI and call it

Comment (SiriusNEO) [reply python/tvm/relax/op/_op_gradient.py]: There are some problems about exposing it to python. Let's just keep it for now

File: python/tvm/relax/op/base.py [hunk 14]
```diff
@@ -414,3 +414,20 @@ def shape_of(expr: Expr) -> Expr:
...
+def register_gradient(
+    op_name: str, fgradient: Callable[[Call, Var], List[Expr]] = None, level: int = 10
+):
+    """Register operator gradient function for a relax operator.
+
+    Parameters
+    ----------
+    op_name : str
+        The name of the op.
+    fgradient : function (orig_call : Call, output_grad : Var) -> new_expr : Expr
```
Comment (MasterJH5574) [inline python/tvm/relax/op/base.py]: The annotation need a bit update

File: python/tvm/relax/op/_op_gradient.py [hunk 15]
```diff
@@ -0,0 +1,491 @@
...
+        y = relax.sum(x, axis, keepdims)
+
+    Backward:
+        Returns [broadcast_to(y_output_grad, x.shape)].
+        If `keepdims=False`, the summed axis will be added back.
+    """
+    axis = orig_call.attrs["axis"]
+    keepdims = orig_call.attrs["keepdims"]
+    if not keepdims and axis:
+        output_grad = expand_dims(output_grad, [int(ax) for ax in axis])
```
Comment (MasterJH5574) [inline python/tvm/relax/op/_op_gradient.py]: I assume directly passing `axis` already works?

File: python/tvm/relax/op/_op_gradient.py [hunk 16]
```diff
@@ -0,0 +1,491 @@
...
+        b_expand = expand_dims(tensor_b, 0)
+        grad_expand = expand_dims(output_grad, -1)
+        a_grad = matmul(grad_expand, b_expand)
+        b_grad = mean(
+            matmul(_transpose_last_two_dim(tensor_a, a_dim), grad_expand), axis=-1
+        )  # squeeze last dim
+        # legalizer now not support squeeze
+        # squeeze(
+        #     matmul(_transpose_last_two_dim(tensor_a, a_dim), output_grad_expand), axis=-1
+        # )
```
Comment (MasterJH5574) [inline python/tvm/relax/op/_op_gradient.py]: Hmmmm what does it mean by not supporting here?

File: tests/python/relax/test_op_gradient_eliminate_collapse.py [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+    op = Op.get(op_name)
+    op_grad_func = op.get_attr("FPrimalGradient")
+    call = op_func(*args)
+    bb = relax.BlockBuilder()
+
+    with bb.function("main", list(args) + [grad_var]):
+        with bb.dataflow():
+            orig_var = bb.emit_output(call)
+            grad_call = relax.Tuple(op_grad_func(orig_var, call, grad_var, bb))
+        bb.emit_func_output(orig_var)
```
Comment (MasterJH5574) [inline tests/python/relax/test_op_gradient_eliminate_collapse.py]: Seems that the generated function is never used? What is the purpose here?

File: tests/python/relax/test_op_gradient_numeric.py [hunk 18]
```diff
@@ -0,0 +1,409 @@
...
+        [data_numpy1, data_numpy2],
+        target,
+        dev,
+        (),
+    )
+
+
+if __name__ == "__main__":
+    # test_permute_dims("llvm", tvm.cpu(0))
+    pytest.main([__file__])
```
Comment (MasterJH5574) [inline tests/python/relax/test_op_gradient_numeric.py]: We can use `tvm.testing.main()`

File: include/tvm/relax/op_attr_types.h [hunk 19]
```diff
@@ -49,6 +49,16 @@ using FInferStructInfo =
...
 
+/*!
+ * \brief Gradient for a specific op.
+ *
+ * \param orig_call the original Expr.
+ * \param output_grad the gradient of the Expr.
+ * \return the gradient for each parameter.
+ */
+using FPrimalGradient = runtime::TypedPackedFunc<tvm::Array<Expr>(
+    const Var& orig_var, const Call& orig_call, const Var& output_grad, BlockBuilder ctx)>;
```
Comment (MasterJH5574) [inline include/tvm/relax/op_attr_types.h]: Maybe `const BlockBuilder&`?

File: include/tvm/relax/op_attr_types.h [hunk 20]
```diff
@@ -49,6 +49,16 @@ using FInferStructInfo =
  */
 using FCallPacked = String;
 
+/*!
+ * \brief Gradient for a specific op.
+ *
+ * \param orig_call the original Expr.
+ * \param output_grad the gradient of the Expr.
```
Comment (MasterJH5574) [inline include/tvm/relax/op_attr_types.h]: Document for params needs update.

File: python/tvm/relax/op/_op_gradient.py (hunk 16 above)
Comment (SiriusNEO) [reply python/tvm/relax/op/_op_gradient.py]: My fault. It is not supported in the previous version of legalizer. Let me update here.

File: tests/python/relax/test_op_gradient_eliminate_collapse.py (hunk 17 above)
Comment (SiriusNEO) [reply tests/python/relax/test_op_gradient_eliminate_collapse.py]: Note that the gradient function need orig_var, and seems that creating such var and binding it to `call` must be done in a context of block builder

File: include/tvm/relax/op_attr_types.h (hunk 19 above)
Comment (SiriusNEO) [reply include/tvm/relax/op_attr_types.h]: I'm not sure since we may use `ctx.emit`. Is it still const?

Comment (MasterJH5574) [reply include/tvm/relax/op_attr_types.h]: Yes this ObjectRef still points to the BlockBuilderNode.

File: tests/python/relax/test_op_gradient_eliminate_collapse.py (hunk 17 above)
Comment (MasterJH5574) [reply tests/python/relax/test_op_gradient_eliminate_collapse.py]: I see. Seems creating a Function is too much. What about
* normalizing the Call by block builder
* creating a var
* updating the struct info of the var using `var._update_struct_info` https://github.com/mlc-ai/relax/blob/760dbe8a71361654fa64fa5f2d8a31f777f7d1ac/python/tvm/relax/expr.py#L684-L685

We don’t have to emit it, as here we only want a Var with the correct sinfo.

File: tests/python/relax/test_op_gradient_numeric.py [hunk 21]
```diff
@@ -0,0 +1,408 @@
...
+    """Generate module and run it to check numberic gradients."""
+
+    func_name = "main"
+
+    # prepare input
+    def _numpy_to_var(data, var_name):
+        if isinstance(data, list):
+            struct_infos = []
+            for i in range(len(data)):
+                tvm_var = _numpy_to_var(data[i], "")
```
Comment (MasterJH5574) [inline tests/python/relax/test_op_gradient_numeric.py]: Just some easier way
```suggestion
            for _data in data:
                tvm_var = _numpy_to_var(_data, "")
```

Comment (MasterJH5574) [reply tests/python/relax/test_op_gradient_numeric.py]: Ditto for those enumerations below.

File: tests/python/relax/test_op_gradient_numeric.py [hunk 22]
```diff
@@ -0,0 +1,408 @@
...
+    def _gen_weights(shape):
+        if isinstance(shape, list):
+            ret = []
+            for s in shape:
+                ret.append(_gen_weights(s))
+            return ret
+        else:
+            return np.random.uniform(size=shape).astype(np.float32)
+
+    param_vars = [_numpy_to_var(inputs_numpy[i], "x_" + str(i)) for i in range(len(inputs_numpy))]
```
Comment (MasterJH5574) [inline tests/python/relax/test_op_gradient_numeric.py]: I think `enumerate` is helpful here.
```suggestion
    param_vars = [_numpy_to_var(input_numpy, "x_" + str(i)) for i, input_numpy in enumerate(inputs_numpy)]
```

**Review by MasterJH5574 — APPROVED** [review]
